### PR TITLE
JSON: Minor improvements

### PR DESCRIPTION
**Pull request by obastemur**
- fast path for constant propertyId
- CompleteInit, do not double copy

### CODE REVIEW

**Comment by obastemur** [comment]
@kunalspathak Thanks for the review!

File: lib/Runtime/Library/JSON.cpp [hunk 1]
```diff
@@ -336,17 +332,28 @@ namespace JSON
...
-        wmemset(buffer, _u(' '), JSONspaceSize);
+        static const char16* spaceBuffer = GetGapStringInternal(); // multi threading safe
+
         charcount_t len = 0;
         switch (Js::JavascriptOperators::GetTypeId(space))
         {
         case Js::TypeIds_Integer:
             {
                 len = max(0, min(JSONspaceSize, static_cast<int>(Js::TaggedInt::ToInt32(space))));
+                if (len)
```
Comment (akroshg) [inline lib/Runtime/Library/JSON.cpp]: nit: can you use len> 0 ?

File: lib/Runtime/Base/ScriptContext.h [hunk 2]
```diff
@@ -1424,7 +1424,20 @@ namespace Js
...
         }
 
+    private:
+        Js::PropertyId emptyStringPropertyId;
     public:
+        Js::PropertyId GetEmptyStringPropertyId()
+        {
+            if (emptyStringPropertyId == Js::PropertyIds::_none)
+            {
+                Js::PropertyRecord const * propertyRecord;
```
Comment (obastemur) [inline lib/Runtime/Base/ScriptContext.h]: why?

File: lib/Runtime/Library/JSON.cpp (hunk 1 above)
Comment (obastemur) [inline lib/Runtime/Library/JSON.cpp]: `max(0, ....)`

File: lib/Runtime/Base/ScriptContext.h (hunk 2 above)
Comment (akroshg) [inline lib/Runtime/Base/ScriptContext.h]: kind of good practice in general. what if for Example GetOrAddPropertyRecord not populate the out variable in all path.

---
In reply to: [131467416](https://github.com/Microsoft/ChakraCore/pull/3455#discussion_r131467416) [](ancestors = 131467416)

File: lib/Runtime/Library/JSON.cpp (hunk 1 above)
Comment (akroshg) [inline lib/Runtime/Library/JSON.cpp]: my point is you are using int as bool expression in the if condition.

---
In reply to: [131467627](https://github.com/Microsoft/ChakraCore/pull/3455#discussion_r131467627) [](ancestors = 131467627)

**Pull request merged**